base-files: /etc/profile service improvement

stangri opened the pull request:
This patch introduces: 
1. ```service enabled``` to list the status (enabled/disabled) of services in ```/etc/init.d```
2. ```service <service-name-here> ubus``` to display the ```ubus call service list``` output for specified service

Signed-off-by: Stan Grishin <stangri@melmac.net>

stangri committed:
base-files: /etc/profile service improvement

This patch introduces:

1. service enabled to list the status (enabled/disabled) of services in /etc/init.d
2. service <service-name-here> ubus to display the ubus call service list output for specified service

Signed-off-by: Stan Grishin <stangri@melmac.net>

adrianschmutzler commented:
Please add the description to the commit message.

stangri commented:
> Please add the description to the commit message.

done.

adrianschmutzler commented:
Description has to be _before_ signed-off, so signed-off is the last line.

ysc3839 commented:
Why not implement `ubus` command in `/etc/rc.common`?

stangri commented:
> Description has to be _before_ signed-off, so signed-off is the last line.

Done.
PS. If it's that important, I think it should be added to the submitting patches text, rather than hotlinking that instruction.

stangri commented:
> Why not implement `ubus` command in `/etc/rc.common`?

I'm not sure if that's the best approach.

ynezz commented on package/base-files/files/etc/profile:
```diff
@@ -49,11 +49,18 @@ in order to prevent unauthorized SSH logins.
-service() {
-	[ -f "/etc/init.d/$1" ] || {
+service() { 
+	if [ "$1" = "enabled" ] && [ -z "$2" ]; then 
```
Seems confusing to me, this is going to be called as `service enabled` and list enabled AND disabled services. 

Perhaps something like `service show-enabled` (or `list-enabled`) and `service show-disabled` (or `list-disabled`) would make more sense?

stangri replied:
Oh, great idea, perhaps list-enabled to be consistent with opkg's list-installed?

Unless there are other comments, I'll implement the fix and push it later this week.

stangri replied:
Implemented list-enabled and list-disabled as per our conversation.

stangri commented:
Implemented the list-enabled and list-disabled as per our earlier conversation. Should I make a separate commit for 19.07 an 18.06?

adrianschmutzler commented:
@stangri Let's focus on master for know. To me this looks like a backport would be trivial, so we wouldn't need a commit for it.

The question to me is whether this should be backported at all, as it looks like a feature to me (rather than a fix).

ynezz commented:
> whether this should be backported

Nope.

wulfy23 commented on package/base-files/files/etc/profile:
```diff
@@ -49,11 +49,22 @@ in order to prevent unauthorized SSH logins.
+	elif [ -f "/etc/init.d/$1" ]; then 
+		/etc/init.d/$@
+	else 
 		echo "service "'"'"$1"'"'" not found, the following services are available:"
```
echo "service list-enabled|list-disabled|<name start|stop|restart|reload|enable|disable|ubus>" ... etc ...

stangri replied:
There could be more commands beyond listed, so my fix is slightly different from suggested.

stangri commented:
> The question to me is whether this should be backported at all, as it looks like a feature to me (rather than a fix).

19.07 hasn't been released yet, I would strongly suggest it's backported to 19.07.

stangri commented:
@adrianschmutzler @ynezz -- Should the "needs changes" tag now be removed? What's preventing this from being merged?

stangri commented:
@wulfy23 @ysc3839 any more feedback?

oldium commented:
Regarding ubus argument - from the architecture point of view simulating/stealing ubus argument in context, which is usually just forwarded to init script, is not good. For me logical would be to be able to call /etc/init.d/someservice ubus, so I would implement it in rc.common too.

stangri commented:
> Regarding ubus argument - from the architecture point of view simulating/stealing ubus argument in context, which is usually just forwarded to init script, is not good. For me logical would be to be able to call /etc/init.d/someservice ubus, so I would implement it in rc.common too.

Moving that ubus call to /etc/rc.common leads to an indefinite loop if I do this:
```sh
ubus() {
	name="$(basename "${initscript}")"
#	echo "ubus - $name"
	ubus call service list "{'name': '$name'}"
	return 0
}

ALL_COMMANDS="start stop reload restart boot shutdown enable disable enabled depends ubus ${EXTRA_COMMANDS}"
list_contains ALL_COMMANDS "$action" || action=help
$action "$@"
```

If you have suggestions on how to overcome this, please let me know.

oldium commented:
`ubus` is not very intuitive anyway, so I would rename `ubus` to something like `status-ubus` or enhance the current `status` by optional parameter `--ubus`.

stangri commented:
> `ubus` is not very intuitive anyway, so I would rename `ubus` command to something like `status-ubus` or enhance the current `status` by optional parameter `--ubus`.

Sure, I can rename it to ubusinfo or whatever. The issue is that calling ubus from rc.common causes the loop. You know how to overcome this?

oldium commented:
Sorry, just updated my latest comment - please see update.

oldium commented:
@stangri Issue with loop is caused by the fact that you have function `ubus()`, which is called in the line `ubus call ...`. When you rename `ubus()` to something else, the loop disappears.

stangri commented:
> Sorry, just updated my latest comment - please see update.

We're on the same page with the name and I can address it later. 

> @stangri Issue with loop is caused by the fact that you have function `ubus()`, which is called in the line `ubus call ...`. When you rename `ubus()` to something else, the loop disappears.

OMG, my only excuse that it's 4:30am on the NY Day! Just replaced the call with `/bin/ubus` and the problem went away.

stangri commented:
Renamed `ubus` to `status-ubus` and pushed fixes to `/etc/profile` and `/etc/rc.common`. I still think `ubus` is a better name as it would have allowed to keep the same tabbing in the `help()` function.

stangri committed:
Merge branch 'master' into base-files-etc-profile-service-improvement

stangri commented:
Any feedback?

ynezz commented:
Please rebase properly and `git push -f` in order to get rid of that additional merge commit.

ynezz commented:
While at it line wrap your commit description to < 72 chars.

stangri commented:
Thanks, I'll submit thru mail list as per your instructions.
